lib/liquid/context.rb

```diff
@@ -191,15 +191,12 @@ def evaluate(object)
     def find_variable(key, raise_on_not_found: true)
       # This was changed from find() to find_index() because this is a very hot
       # path and find_index() is optimized in MRI to reduce object allocation
-      index = @scopes.find_index { |s| s.key?(key) }
+      scope = (index = @scopes.find_index { |s| s.key?(key) }) && @scopes[index]
+      scope ||= (index = @environments.find_index { |s| s.key?(key) || !s.default_proc.nil? }) && @environments[index]
+      scope ||= (index = @static_environments.find_index { |s| s.key?(key) }) && @static_environments[index]
+      scope ||= {}
 
-      variable = if index
-        lookup_and_evaluate(@scopes[index], key, raise_on_not_found: raise_on_not_found)
-      else
-        try_variable_find_in_environments(key, raise_on_not_found: raise_on_not_found)
-      end
-
-      variable = variable.to_liquid
+      variable = lookup_and_evaluate(scope, key, raise_on_not_found: raise_on_not_found).to_liquid
```

Comment:
Can we remove `raise_on_not_found` now and do that work in this method? i.e.

```rb
      scope = (index = @scopes.find_index { |s| s.key?(key) }) && @scopes[index]
      scope ||= (index = @environments.find_index { |s| s.key?(key) || !s.default_proc.nil? }) && @environments[index]
      scope ||= (index = @static_environments.find_index { |s| s.key?(key) }) && @static_environments[index]

      unless scope
        if @strict_variables
          raise Liquid::UndefinedVariable, "undefined variable #{key}"
        else
          return nil
        end
      end

      lookup_and_evaluate(scope, key).to_liquid
```

Reply:
I do like this as it shortcuts the response

Reply:
Only downfall is it does duplicate the code as VariableLookup is also needing to run this anyway

Reply:
Though it doesn't use `raise_on_not_found`, so that logic could still be moved right?